// ng-otp-input/bundles/ng-otp-input.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-otp-input', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ng-otp-input'] = {}, global.ng.core, global.ng.forms, global.ng.common));
}(this, (function (exports, core, forms, common) { 'use strict';

    var KeysPipe = /** @class */ (function () {
        function KeysPipe() {
        }
        KeysPipe.prototype.transform = function (value) {
            return Object.keys(value);
        };
        return KeysPipe;
    }());
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    KeysPipe.decorators = [
        { type: core.Pipe, args: [{
                    name: 'keys'
                },] }
    ];

    var NgOtpInputComponent = /** @class */ (function () {
        function NgOtpInputComponent(keysPipe) {
            this.keysPipe = keysPipe;
            this.config = { length: 4 };
            // tslint:disable-next-line: no-output-on-prefix
            this.onInputChange = new core.EventEmitter();
            this.inputControls = new Array(this.config.length);
            this.componentKey = Math.random()
                .toString(36)
                .substring(2) + new Date().getTime().toString(36);
        }
        NgOtpInputComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.otpForm = new forms.FormGroup({});
            for (var index = 0; index < this.config.length; index++) {
                this.otpForm.addControl(this.getControlName(index), new forms.FormControl());
            }
            this.otpForm.valueChanges.subscribe(function (v) {
                _this.keysPipe.transform(_this.otpForm.controls).forEach(function (k) {
                    var val = _this.otpForm.controls[k].value;
                    if (val && val.length > 1) {
                        if (val.length >= _this.config.length) {
                            _this.setValue(val);
                        }
                        else {
                            _this.rebuildValue();
                        }
                    }
                });
            });
            this.inputType = this.getInputType();
        };
        NgOtpInputComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            if (!this.config.disableAutoFocus) {
                var containerItem = document.getElementById("c_" + this.componentKey);
                if (containerItem) {
                    containerItem.addEventListener('paste', function (evt) { return _this.handlePaste(evt); });
                    var ele = containerItem.getElementsByClassName('otp-input')[0];
                    if (ele && ele.focus) {
                        ele.focus();
                    }
                }
            }
        };
        NgOtpInputComponent.prototype.getControlName = function (idx) {
            return "ctrl_" + idx;
        };
        NgOtpInputComponent.prototype.ifLeftArrow = function (event) {
            return this.ifKeyCode(event, 37);
        };
        NgOtpInputComponent.prototype.ifRightArrow = function (event) {
            return this.ifKeyCode(event, 39);
        };
        NgOtpInputComponent.prototype.ifBackspaceOrDelete = function (event) {
            return (event.key === 'Backspace' ||
                event.key === 'Delete' ||
                this.ifKeyCode(event, 8) ||
                this.ifKeyCode(event, 46));
        };
        NgOtpInputComponent.prototype.ifKeyCode = function (event, targetCode) {
            var key = event.keyCode || event.charCode;
            // tslint:disable-next-line: triple-equals
            return key == targetCode ? true : false;
        };
        NgOtpInputComponent.prototype.onKeyDown = function ($event) {
            var isSpace = this.ifKeyCode($event, 32);
            if (isSpace) { // prevent space
                return false;
            }
        };
        NgOtpInputComponent.prototype.onKeyUp = function ($event, inputIdx) {
            var nextInputId = this.appendKey("otp_" + (inputIdx + 1));
            var prevInputId = this.appendKey("otp_" + (inputIdx - 1));
            if (this.ifRightArrow($event)) {
                this.setSelected(nextInputId);
                return;
            }
            if (this.ifLeftArrow($event)) {
                this.setSelected(prevInputId);
                return;
            }
            var isBackspace = this.ifBackspaceOrDelete($event);
            if (isBackspace && !$event.target.value) {
                this.setSelected(prevInputId);
                this.rebuildValue();
                return;
            }
            if (!$event.target.value) {
                return;
            }
            if (this.ifValidEntry($event)) {
                this.setSelected(nextInputId);
            }
            this.rebuildValue();
        };
        NgOtpInputComponent.prototype.appendKey = function (id) {
            return id + "_" + this.componentKey;
        };
        NgOtpInputComponent.prototype.setSelected = function (eleId) {
            this.focusTo(eleId);
            var ele = document.getElementById(eleId);
            if (ele && ele.setSelectionRange) {
                setTimeout(function () {
                    ele.setSelectionRange(0, 1);
                }, 0);
            }
        };
        NgOtpInputComponent.prototype.ifValidEntry = function (event) {
            var inp = String.fromCharCode(event.keyCode);
            var isMobile = /iPhone|iPad|iPod|Android/i.test(navigator.userAgent);
            return (isMobile ||
                /[a-zA-Z0-9-_]/.test(inp) ||
                (this.config.allowKeyCodes &&
                    this.config.allowKeyCodes.includes(event.keyCode)) ||
                (event.keyCode >= 96 && event.keyCode <= 105));
        };
        NgOtpInputComponent.prototype.focusTo = function (eleId) {
            var ele = document.getElementById(eleId);
            if (ele) {
                ele.focus();
            }
        };
        // method to set component value
        NgOtpInputComponent.prototype.setValue = function (value) {
            var _this = this;
            if (this.config.allowNumbersOnly && isNaN(value)) {
                return;
            }
            this.otpForm.reset();
            if (!value) {
                this.rebuildValue();
                return;
            }
            value = value.toString().replace(/\s/g, ''); // remove whitespace
            Array.from(value).forEach(function (c, idx) {
                if (_this.otpForm.get(_this.getControlName(idx))) {
                    _this.otpForm.get(_this.getControlName(idx)).setValue(c);
                }
            });
            if (!this.config.disableAutoFocus) {
                var containerItem = document.getElementById("c_" + this.componentKey);
                var indexOfElementToFocus = value.length < this.config.length ? value.length : (this.config.length - 1);
                var ele = containerItem.getElementsByClassName('otp-input')[indexOfElementToFocus];
                if (ele && ele.focus) {
                    ele.focus();
                }
            }
            this.rebuildValue();
        };
        NgOtpInputComponent.prototype.rebuildValue = function () {
            var _this = this;
            var val = '';
            this.keysPipe.transform(this.otpForm.controls).forEach(function (k) {
                if (_this.otpForm.controls[k].value) {
                    var ctrlVal = _this.otpForm.controls[k].value;
                    var isLengthExceed = ctrlVal.length > 1;
                    var isCaseTransformEnabled = !_this.config.allowNumbersOnly && _this.config.letterCase && (_this.config.letterCase.toLocaleLowerCase() == 'upper' || _this.config.letterCase.toLocaleLowerCase() == 'lower');
                    ctrlVal = ctrlVal[0];
                    var transformedVal = isCaseTransformEnabled ? _this.config.letterCase.toLocaleLowerCase() == 'upper' ? ctrlVal.toUpperCase() : ctrlVal.toLowerCase() : ctrlVal;
                    if (isCaseTransformEnabled && transformedVal == ctrlVal) {
                        isCaseTransformEnabled = false;
                    }
                    else {
                        ctrlVal = transformedVal;
                    }
                    val += ctrlVal;
                    if (isLengthExceed || isCaseTransformEnabled) {
                        _this.otpForm.controls[k].setValue(ctrlVal);
                    }
                }
            });
            this.onInputChange.emit(val);
        };
        NgOtpInputComponent.prototype.getInputType = function () {
            return this.config.isPasswordInput
                ? 'password'
                : this.config.allowNumbersOnly
                    ? 'tel'
                    : 'text';
        };
        NgOtpInputComponent.prototype.handlePaste = function (e) {
            // Get pasted data via clipboard API
            var clipboardData = e.clipboardData || window['clipboardData'];
            if (clipboardData) {
                var pastedData = clipboardData.getData('Text');
            }
            // Stop data actually being pasted into div
            e.stopPropagation();
            e.preventDefault();
            if (!pastedData) {
                return;
            }
            this.setValue(pastedData);
        };
        return NgOtpInputComponent;
    }());
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgOtpInputComponent.decorators = [
        { type: core.Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: 'ng-otp-input',
                    template: "<div class=\"wrapper {{config.containerClass}}\" id=\"c_{{componentKey}}\" *ngIf=\"otpForm?.controls\"\r\n  [ngStyle]=\"config.containerStyles\">\r\n  <input [pattern]=\"config.allowNumbersOnly ? '\\\\d*' : ''\" [type]=\"inputType\" numberOnly [placeholder]=\"config?.placeholder || ''\"\r\n    [disabledNumberOnly]=\"!config.allowNumbersOnly\" [ngStyle]=\"config.inputStyles\" \r\n    class=\"otp-input {{config.inputClass}}\" autocomplete=\"one-time-code\" *ngFor=\"let item of otpForm?.controls | keys;let i=index\"\r\n    [formControl]=\"otpForm.controls[item]\" id=\"otp_{{i}}_{{componentKey}}\" (keydown)=\"onKeyDown($event)\"\r\n    (keyup)=\"onKeyUp($event,i)\">\r\n</div>",
                    styles: [".otp-input{width:50px;height:50px;border-radius:4px;border:solid 1px #c5c5c5;text-align:center;font-size:32px}.wrapper .otp-input:not(:last-child){margin-right:8px}@media screen and (max-width: 767px){.otp-input{width:40px;font-size:24px;height:40px}}@media screen and (max-width: 420px){.otp-input{width:30px;font-size:18px;height:30px}}\n"]
                },] }
    ];
    /**
     * @type {function(): !Array<(null|{
     *   type: ?,
     *   decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>),
     * })>}
     * @nocollapse
     */
    NgOtpInputComponent.ctorParameters = function () { return [
        { type: KeysPipe }
    ]; };
    /** @type {!Object<string, !Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NgOtpInputComponent.propDecorators = {
        config: [{ type: core.Input }],
        onInputChange: [{ type: core.Output }]
    };

    var NumberOnlyDirective = /** @class */ (function () {
        function NumberOnlyDirective(_elRef, _renderer) {
            this._elRef = _elRef;
            this._renderer = _renderer;
        }
        NumberOnlyDirective.prototype.ngOnInit = function () {
            if (!this.disabledNumberOnly) {
                this._renderer.setAttribute(this._elRef.nativeElement, 'onkeypress', 'return (event.charCode >= 48 && event.charCode <= 57) || event.charCode == 0');
            }
        };
        return NumberOnlyDirective;
    }());
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NumberOnlyDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[numberOnly]'
                },] }
    ];
    /**
     * @type {function(): !Array<(null|{
     *   type: ?,
     *   decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>),
     * })>}
     * @nocollapse
     */
    NumberOnlyDirective.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.Renderer2 }
    ]; };
    /** @type {!Object<string, !Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    NumberOnlyDirective.propDecorators = {
        disabledNumberOnly: [{ type: core.Input }]
    };

    var NgOtpInputModule = /** @class */ (function () {
        function NgOtpInputModule() {
        }
        return NgOtpInputModule;
    }());
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgOtpInputModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule,
                        forms.FormsModule,
                        forms.ReactiveFormsModule
                    ],
                    declarations: [NgOtpInputComponent, KeysPipe, NumberOnlyDirective],
                    exports: [NgOtpInputComponent],
                    providers: [KeysPipe]
                },] }
    ];

    /*
     * Public API Surface of ng-otp-input
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NgOtpInputComponent = NgOtpInputComponent;
    exports.NgOtpInputModule = NgOtpInputModule;
    exports.ɵa = KeysPipe;
    exports.ɵb = NumberOnlyDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
